File: frontend/src/types/import-source.ts
/**
 * TypeScript types for Import Sources and Profiles (v0.8)
 * Must match backend/app/schemas/import_source.py and import_profile.py
 */

// =============================================================================
// Enums - Must match backend/app/db/models/enums.py
// =============================================================================

// Must match backend/app/db/models/enums.py
export type ImportSourceType = 'GOOGLE_DRIVE' | 'UPLOAD' | 'BULK_FOLDER' | 'PHPBB_FORUM'

export type ImportSourceStatus = 'ACTIVE' | 'PAUSED' | 'ERROR' | 'PENDING' | 'RATE_LIMITED'

export type ImportRecordStatus = 'PENDING' | 'IMPORTING' | 'IMPORTED' | 'SKIPPED' | 'ERROR'

export type ConflictResolution = 'SKIP' | 'REPLACE' | 'ASK'

// =============================================================================
// Import Profile Configuration Types
// =============================================================================

export interface ProfileDetectionConfig {
  model_extensions: string[]
  archive_extensions: string[]
  min_model_files: number
  structure: 'nested' | 'flat' | 'auto'
  model_subfolders: string[]
}

export interface ProfileTitleConfig {
  source: 'folder_name' | 'parent_folder' | 'filename'
  strip_patterns: string[]
  case_transform: 'none' | 'title' | 'lower' | 'upper'
}

export interface ProfilePreviewConfig {
  folders: string[]
  wildcard_folders: string[]
  extensions: string[]
  include_root: boolean
}

export interface ProfileIgnoreConfig {
  folders: string[]
  extensions: string[]
  patterns: string[]
}

export interface ProfileAutoTagConfig {
  from_subfolders: boolean
  from_filename: boolean
  subfolder_levels: number
  strip_patterns: string[]
}

export interface ImportProfileConfig {
  detection: ProfileDetectionConfig
  title: ProfileTitleConfig
  preview: ProfilePreviewConfig
  ignore: ProfileIgnoreConfig
  auto_tags: ProfileAutoTagConfig
}

// =============================================================================
// Import Profile Types
// =============================================================================

export interface ImportProfileSummary {
  id: string
  name: string
  is_builtin: boolean
}

export interface ImportProfile {
  id: string
  name: string
  description: string | null
  is_builtin: boolean
  config: ImportProfileConfig
  created_at: string
  updated_at: string
}

export interface ImportProfileCreate {
  name: string
  description?: string | null
  config?: ImportProfileConfig
}

export interface ImportProfileUpdate {
  name?: string
  description?: string | null
  config?: ImportProfileConfig
}

export interface ImportProfileList {
  items: ImportProfile[]
  total: number
}

export interface ImportProfileUsage {
  sources_using: number
  source_names: string[]
}

// =============================================================================
// Import Source Folder Types (DEC-038)
// =============================================================================

export interface ImportSourceFolderSummary {
  id: string
  name: string | null
  google_drive_url: string | null
  google_folder_id: string | null
  folder_path: string | null
  phpbb_forum_url: string | null // v1.0 - issue #242
  enabled: boolean
  items_detected: number
  items_imported: number
  last_synced_at: string | null
  has_overrides: boolean
}

export interface ImportSourceFolder extends ImportSourceFolderSummary {
  import_source_id: string
  import_profile_id: string | null
  default_designer: string | null
  default_tags: string[] | null
  effective_profile_id: string | null
  effective_designer: string | null
  effective_tags: string[]
  sync_cursor: string | null
  last_sync_error: string | null
  created_at: string
  updated_at: string
}

export interface ImportSourceFolderCreate {
  name?: string | null
  google_drive_url?: string | null
  folder_path?: string | null
  phpbb_forum_url?: string | null // v1.0 - issue #242
  import_profile_id?: string | null
  default_designer?: string | null
  default_tags?: string[] | null
  enabled?: boolean
}

export interface ImportSourceFolderUpdate {
  name?: string | null
  import_profile_id?: string | null
  default_designer?: string | null
  default_tags?: string[] | null
  enabled?: boolean | null
}

// =============================================================================
// Import Source Types
// =============================================================================

export interface ImportSource {
  id: string
  name: string
  source_type: ImportSourceType
  status: ImportSourceStatus

  // Google OAuth status (shared across all folders)
  google_connected?: boolean

  // phpBB connection status (v1.0 - issue #241)
  phpbb_connected?: boolean
  phpbb_forum_url?: string | null

  // Type-specific fields (DEPRECATED - use folders instead)
  google_drive_url: string | null
  google_drive_folder_id: string | null
  folder_path: string | null

  // Settings
  import_profile_id: string | null
  default_designer: string | null
  default_tags: string[] | null
  sync_enabled: boolean
  sync_interval_hours: number

  // State
  last_sync_at: string | null
  last_sync_error: string | null
  items_imported: number

  // Timestamps
  created_at: string
  updated_at: string

  // Embedded profile (optional)
  profile: ImportProfileSummary | null

  // Multi-folder support (DEC-038)
  folder_count: number
  folders: ImportSourceFolderSummary[]
}

export interface ImportSourceDetail extends ImportSource {
  pending_count: number
  imported_count: number
  error_count: number
}

export interface ImportSourceCreate {
  name: string
  source_type: ImportSourceType

  // Google Drive specific
  google_drive_url?: string | null
  google_credentials_id?: string | null

  // Bulk folder specific
  folder_path?: string | null

  // phpBB specific (v1.0 - issue #241)
  phpbb_credentials_id?: string | null
  phpbb_forum_url?: string | null

  // Optional settings
  import_profile_id?: string | null
  default_designer?: string | null
  default_tags?: string[] | null
  sync_enabled?: boolean
  sync_interval_hours?: number
}

export interface ImportSourceUpdate {
  name?: string

  // Type-specific (cannot change source_type)
  google_drive_url?: string | null
  folder_path?: string | null

  // Settings
  import_profile_id?: string | null
  default_designer?: string | null
  default_tags?: string[] | null
  sync_enabled?: boolean | null
  sync_interval_hours?: number | null
}

export interface ImportSourceList {
  items: ImportSource[]
  total: number
}

export interface ImportSourceListParams {
  source_type?: ImportSourceType
  sync_enabled?: boolean
  status?: ImportSourceStatus
  page?: number
  page_size?: number
}

// =============================================================================
// Sync & History Types
// =============================================================================

export interface SyncTriggerRequest {
  conflict_resolution?: ConflictResolution
  auto_import?: boolean
}

export interface SyncTriggerResponse {
  source_id: string
  job_id: string | null
  message: string
  designs_detected: number
  designs_imported: number
}

export interface ImportHistoryItem {
  id: string
  source_path: string
  status: ImportRecordStatus
  detected_title: string | null
  design_id: string | null
  error_message: string | null
  detected_at: string
  imported_at: string | null
}

export interface ImportHistoryResponse {
  items: ImportHistoryItem[]
  total: number
  page: number
  page_size: number
}

export interface ImportHistoryParams {
  status?: ImportRecordStatus
  page?: number
  page_size?: number
}

// =============================================================================
// Google OAuth Types
// =============================================================================

export interface GoogleOAuthStatus {
  configured: boolean
  api_key_configured: boolean
  authenticated: boolean
  email: string | null
  expires_at: string | null
}

export interface GoogleCredentials {
  id: string
  email: string
  expires_at: string | null
  created_at: string
  updated_at: string
}

export interface GoogleCredentialsList {
  items: GoogleCredentials[]
  total: number
}

export interface GoogleOAuthInitResponse {
  authorization_url: string
  state: string
}

export interface GoogleOAuthCallbackParams {
  code: string
  state: string
}

export interface GoogleOAuthCallbackResponse {
  success: boolean
  credentials_id: string
  email: string
  message: string
}

// =============================================================================
// phpBB Credentials Types (v1.0 - issue #241)
// =============================================================================

export interface PhpbbTestLoginRequest {
  base_url: string
  username: string
  password: string
}

export interface PhpbbTestLoginResponse {
  success: boolean
  message: string
}

export interface PhpbbCredentialsCreate {
  base_url: string
  username: string
  password: string
  test_login?: boolean
}

export interface PhpbbCredentials {
  id: string
  base_url: string
  last_login_at: string | null
  last_login_error: string | null
  session_expires_at: string | null
  created_at: string
}

export interface PhpbbCredentialsList {
  items: PhpbbCredentials[]
}

File: backend/app/schemas/import_source.py
"""Pydantic schemas for Import Source API (v0.8).

Updated for DEC-038 multi-folder support.
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

from app.db.models.enums import ConflictResolution, ImportSourceStatus, ImportSourceType


# ============================================================
# Folder Schemas (DEC-038)
# ============================================================


class ImportSourceFolderCreate(BaseModel):
    """Schema for adding a folder to an import source."""

    name: str | None = Field(
        None, max_length=255, description="Optional display name for the folder"
    )

    # Location (one required based on source type)
    google_drive_url: str | None = Field(
        None, max_length=1024, description="Google Drive folder URL"
    )
    folder_path: str | None = Field(
        None, max_length=1024, description="Local folder path"
    )

    # Per-folder overrides (optional - inherit from source if not set)
    import_profile_id: str | None = Field(
        None, description="Override source's import profile"
    )
    default_designer: str | None = Field(
        None, max_length=255, description="Override source's default designer"
    )
    default_tags: list[str] | None = Field(
        None, description="Override source's default tags"
    )

    enabled: bool = Field(True, description="Whether folder is enabled for sync")


class ImportSourceFolderUpdate(BaseModel):
    """Schema for updating a folder (all fields optional)."""

    name: str | None = Field(None, max_length=255)

    # Location cannot be changed - delete and recreate instead

    # Per-folder overrides
    import_profile_id: str | None = None
    default_designer: str | None = Field(None, max_length=255)
    default_tags: list[str] | None = None

    enabled: bool | None = None


class ImportSourceFolderSummary(BaseModel):
    """Summary of a folder for embedding in source responses."""

    model_config = {"from_attributes": True}

    id: str
    name: str | None = None
    google_drive_url: str | None = None
    google_folder_id: str | None = None
    folder_path: str | None = None
    enabled: bool = True
    items_detected: int = 0
    items_imported: int = 0
    last_synced_at: datetime | None = None
    has_overrides: bool = False  # True if any override is set


class ImportSourceFolderResponse(BaseModel):
    """Full folder response with all details."""

    model_config = {"from_attributes": True}

    id: str
    import_source_id: str
    name: str | None = None

    # Location
    google_drive_url: str | None = None
    google_folder_id: str | None = None
    folder_path: str | None = None

    # Overrides (null = inherit from source)
    import_profile_id: str | None = None
    default_designer: str | None = None
    default_tags: list[str] | None = None

    # Effective values (computed from source + overrides)
    effective_profile_id: str | None = None
    effective_designer: str | None = None
    effective_tags: list[str] = []

    # State
    enabled: bool = True
    last_synced_at: datetime | None = None
    sync_cursor: str | None = None
    last_sync_error: str | None = None

    # Stats
    items_detected: int = 0
    items_imported: int = 0

    # Timestamps
    created_at: datetime


class FolderSyncTriggerResponse(BaseModel):
    """Response after triggering sync for a single folder."""

    folder_id: str
    job_id: str | None = None
    message: str
    designs_detected: int = 0
    designs_imported: int = 0


# ============================================================
# Source Schemas (original, with folder support)
# ============================================================


class ImportSourceCreate(BaseModel):
    """Schema for creating a new import source."""

    name: str = Field(..., min_length=1, max_length=255)
    source_type: ImportSourceType

    # Google Drive specific
    google_drive_url: str | None = Field(
        None, max_length=1024, description="Google Drive folder/file URL"
    )
    google_credentials_id: str | None = Field(
        None, description="ID of stored Google OAuth credentials for private folders"
    )

    # Bulk folder specific
    folder_path: str | None = Field(
        None, max_length=1024, description="Local folder path"
    )

    # phpBB Forum specific (v1.0 - issue #239)
    phpbb_credentials_id: str | None = Field(
        None, description="ID of stored phpBB credentials"
    )
    phpbb_forum_url: str | None = Field(
        None, max_length=1024, description="URL of the phpBB forum to import from (viewforum.php?f=X)"
    )

    # Optional settings
    import_profile_id: str | None = Field(None, description="Import profile to use")
    default_designer: str | None = Field(
        None, max_length=255, description="Default designer for imported designs"
    )
    default_tags: list[str] | None = Field(
        None, description="Default tags for imported designs"
    )
    sync_enabled: bool = Field(True, description="Enable automatic syncing")
    sync_interval_hours: int = Field(
        1, ge=1, le=168, description="Hours between sync checks"
    )


class ImportSourceUpdate(BaseModel):
    """Schema for updating an import source (all fields optional)."""

    name: str | None = Field(None, min_length=1, max_length=255)

    # Note: source_type cannot be changed after creation

    # Google Drive specific
    google_drive_url: str | None = Field(None, max_length=1024)

    # Bulk folder specific
    folder_path: str | None = Field(None, max_length=1024)

    # Optional settings
    import_profile_id: str | None = None
    default_designer: str | None = Field(None, max_length=255)
    default_tags: list[str] | None = None
    sync_enabled: bool | None = None
    sync_interval_hours: int | None = Field(None, ge=1, le=168)


class ImportProfileSummary(BaseModel):
    """Summary of an import profile for embedding in source responses."""

    id: str
    name: str
    is_builtin: bool


class ImportSourceResponse(BaseModel):
    """Schema for import source response."""

    model_config = {"from_attributes": True}

    id: str
    name: str
    source_type: ImportSourceType
    status: ImportSourceStatus

    # Google OAuth status (shared across all folders)
    google_connected: bool = False

    # phpBB status (v1.0 - issue #239)
    phpbb_connected: bool = False
    phpbb_forum_url: str | None = None

    # Type-specific fields (DEPRECATED - use folders instead)
    google_drive_url: str | None = None
    google_drive_folder_id: str | None = None
    folder_path: str | None = None

    # Shared settings
    import_profile_id: str | None = None
    default_designer: str | None = None
    default_tags: list[str] | None = None
    sync_enabled: bool
    sync_interval_hours: int

    # State
    last_sync_at: datetime | None = None
    last_sync_error: str | None = None
    items_imported: int = 0

    # Folder info (DEC-038)
    folder_count: int = 0
    enabled_folder_count: int = 0
    folders: list[ImportSourceFolderSummary] = []

    # Timestamps
    created_at: datetime
    updated_at: datetime

    # Embedded profile (optional)
    profile: ImportProfileSummary | None = None


class ImportSourceList(BaseModel):
    """Schema for paginated import source list response."""

    items: list[ImportSourceResponse]
    total: int


class ImportSourceDetailResponse(ImportSourceResponse):
    """Detailed import source response with additional info."""

    # Include import records summary
    pending_count: int = 0
    imported_count: int = 0
    error_count: int = 0


class SyncTriggerRequest(BaseModel):
    """Request to trigger a sync for an import source."""

    conflict_resolution: ConflictResolution = Field(
        ConflictResolution.SKIP, description="How to handle conflicts"
    )
    auto_import: bool = Field(
        False, description="Automatically import detected designs"
    )


class SyncTriggerResponse(BaseModel):
    """Response after triggering a sync."""

    source_id: str
    job_id: str | None = None  # If async job was created
    message: str
    designs_detected: int = 0
    designs_imported: int = 0


class ImportHistoryItem(BaseModel):
    """Single item in import history."""

    id: str
    source_path: str
    status: str
    detected_title: str | None = None
    design_id: str | None = None
    error_message: str | None = None
    detected_at: datetime
    imported_at: datetime | None = None


class ImportHistoryResponse(BaseModel):
    """Paginated import history response."""

    items: list[ImportHistoryItem]
    total: int
    page: int
    page_size: int


class GoogleOAuthStatusResponse(BaseModel):
    """Status of Google OAuth authentication."""

    configured: bool
    authenticated: bool
    email: str | None = None
    expires_at: datetime | None = None


# ============================================================
# phpBB Forum Schemas (v1.0 - issue #239)
# ============================================================


class PhpbbCredentialsCreate(BaseModel):
    """Schema for creating phpBB forum credentials."""

    base_url: str = Field(
        ..., max_length=512, description="Base URL of the phpBB forum (e.g., https://hex3dpatreon.com)"
    )
    username: str = Field(..., min_length=1, max_length=255, description="Forum username")
    password: str = Field(..., min_length=1, description="Forum password")
    test_login: bool = Field(True, description="Test login before storing credentials")


class PhpbbCredentialsResponse(BaseModel):
    """Schema for phpBB credentials response (excludes sensitive data)."""

    model_config = {"from_attributes": True}

    id: str
    base_url: str
    last_login_at: datetime | None = None
    last_login_error: str | None = None
    session_expires_at: datetime | None = None
    created_at: datetime


class PhpbbCredentialsList(BaseModel):
    """List of phpBB credentials."""

    items: list[PhpbbCredentialsResponse]


class PhpbbTestLoginRequest(BaseModel):
    """Request to test phpBB login."""

    base_url: str = Field(..., max_length=512, description="Base URL of the phpBB forum")
    username: str = Field(..., min_length=1, max_length=255, description="Forum username")
    password: str = Field(..., min_length=1, description="Forum password")


class PhpbbTestLoginResponse(BaseModel):
    """Response from phpBB login test."""

    success: bool
    message: str
    forum_name: str | None = None


class PhpbbForumInfo(BaseModel):
    """Information about a phpBB forum."""

    forum_id: int
    name: str
    url: str
    topic_count: int = 0


class PhpbbForumListRequest(BaseModel):
    """Request to list forums on a phpBB site."""

    credentials_id: str = Field(..., description="ID of stored phpBB credentials")


class PhpbbForumListResponse(BaseModel):
    """Response listing available forums."""

    forums: list[PhpbbForumInfo]
